File: Dynamic Programming/MaxSubsetSumNoAdjacent.py
# Max Subset Sum No Adjacent

# SOLUTION 2
# O(n) time | O(1) space
def maxSubsetSumNoAdjacent(array):
	if not len(array):
		return 0
	elif len(array) == 1:
		return array[0]
	first = array[0]
	second = max(first, array[1])
	for i in range(2, len(array)):
		cur_max = max(second, first + array[i])
		first = second
		second = cur_max	
	return second



# SOLUTION 1
# O(n) time | O(n) space
def maxSubsetSumNoAdjacent(array):
    # # Write your code here.
    # # pass
	# # Check if array is empty or only has 1 item
	# 	# Return 0 or [0] respectively
	# # Iterate through array[2:-1]
	# # 
	
	# if not len(array):
	# 	return 0
	# elif len(array) == 1:
	# 	return array[0]
	# max_sum = array[:]
	# max_sum[1] = max(max_sum[0], max_sum[1])
	# for i in range(2, len(array)):
	# 	max_sum[i] = max(max_sum[i-1], max_sum[i-2] + array[i])
	
	# return max_sum[-1]

File: Dynamic Programming/MaxSubsetSumNoAdjacent.java
import java.util.*;


// SOLUTION 2: With Space optimization
class Program {
	// O(n) Time | O(1) Space
  public static int maxSubsetSumNoAdjacent(int[] array) {
		if(array.length == 0)
			return 0;
		else if (array.length ==1)
			return array[0];
		
		int iMin2 = array[0];
		int iMin1 = Math.max(array[0], array[1]);
			
		for (int i = 2; i < array.length; i++) {
			int tmp = Math.max(iMin1, iMin2 + array[i]);
			iMin2 = iMin1;
			iMin1 = tmp;
		}
		return iMin1;
  }
}



// SOLUTION 1
class Program {
	// O(n) Time | O(n) space
  public static int maxSubsetSumNoAdjacent(int[] array) {
    int n = array.length;
		if ( n == 0)
			return 0;
		if (n == 1)
			return array[0];
		
		int[] dp = new int[n];
		dp[0] = array[0];
		dp[1] = Math.max(dp[0], array[1]);
		
		for (int i = 2; i < n; i++) {
			dp[i] = Math.max(dp[i - 1], dp[i - 2] + array[i]);
		}
    return dp[n-1];
  }
}
